### PR TITLE
Handle errors when Dodona has no push rights to exercise repo

### LINKED ISSUE #3298: Be clearer that Dodona needs push rights to exercise repositories
At least [this translation](https://github.com/dodona-edu/dodona/blob/develop/config/locales/views/repositories/nl.yml#L4) needs to be updated, but if we could somehow check at creation time that we have the correct rights and refuse to create if not that would be optimal I think.

### PR DESCRIPTION
**Pull request by BramDevlaminck**
This pull request gives an error when Dodona has no push rights to an exercise repo.
This way we are notified when pushing fails

This is how the email looks like (in Dutch) when pushing fails

html:
image

plain text:
image

Closes #3298  .

### CODE REVIEW

**Review by chvp — CHANGES_REQUESTED** [review]
[This translation](https://github.com/dodona-edu/dodona/blob/develop/config/locales/views/repositories/nl.yml#L4) should be updated to mention the push requirement (also in English).

File: app/models/repository.rb [hunk 1]
```diff
@@ -192,7 +192,12 @@ def process_activities
       c['internals']['_info'] = 'These fields are used for internal bookkeeping in Dodona, please do not change them.'
       act.config_file.write(JSON.pretty_generate(c))
     end
-    commit 'stored tokens in new activities' unless new_activities.empty?
+
+    unless new_activities.empty?
+      status, err = commit 'stored tokens in new activities'
+      # handle errors when commit fails
+      errors.push err unless status
```
[the remark this answers is not in the document]
Comment (BramDevlaminck) [reply app/models/repository.rb:199]: Ah, in this case I assume I have to do push a new ConfigParseError object to this array. Or should I not use this type of error to process the error message given by git?

[the remark this answers is not in the document]
Comment (BramDevlaminck) [reply app/models/repository.rb]: I've updated the pr description with how the email looks like

File: test/models/repository_test.rb [hunk 2]
```diff
@@ -304,4 +304,20 @@ def teardown
     @echo.reload
     assert_equal 'not_valid', @echo.status
   end
+
+  test 'should catch error when commit fails' do
```
Comment (jorg-vr) [inline test/models/repository_test.rb:308]: Could you add another test that checks if the email gets send?

Comment (BramDevlaminck) [reply test/models/repository_test.rb:308]: Yes, I'll do

File: config/locales/views/error_mailer/nl.yml [hunk 3]
```diff
@@ -10,3 +10,15 @@ nl:
         other: 'Bij het automatisch verwerken van de oefeningenrepository "%{repository}" liep iets fout. Vermoedelijk zit er een syntaxfout in het configuratiebestand. Onderstaande foutmeldingen werden gegenereerd:'
       regards: 'Vriendelijke groeten,'
       auto-generated: ps. Dit bericht werd automatisch gegenereerd.
+    git_error:
+      subject: 'Pushen naar "%{repository}" mislukt'
+      body:
+        greeting: |
+          Beste %{name},
+        error_message_html: |
+          Tijdens het verwerken "%{repository}" kon Dodona de aanpassingen van de verwerkte oefeningen niet pushen. De exacte error was: %{error}.
```
Comment (chvp) [inline config/locales/views/error_mailer/nl.yml:19]:
```suggestion
          Tijdens het verwerken van "%{repository}" kon Dodona de aanpassingen aan de verwerkte oefeningen niet pushen. De exacte error was: %{error}.
```

File: config/locales/views/error_mailer/nl.yml [hunk 4]
```diff
@@ -10,3 +10,15 @@ nl:
...
+
+          Zorg er zeker voor dat de Dodona gebruiker push rechten heeft op je repository.
```
Comment (chvp) [inline config/locales/views/error_mailer/nl.yml:21]:
```suggestion
          Zorg er zeker voor dat de Dodona gebruiker pushrechten heeft op je repository.
```

File: config/locales/views/error_mailer/en.yml [hunk 5]
```diff
@@ -10,3 +10,15 @@ en:
         other: 'Something went wrong during the processing of the exercise repository "%{repository}". You probably made a syntax error in the config file. These were the generated error messages:'
       regards: "Kind regards,"
       auto-generated: This message was generated automatically.
+    git_error:
+      subject: 'Failed to push to "%{repository}"'
+      body:
+        greeting: |
+          Dear %{name},
+        error_message_html: |
+          While processing "%{repository}", Dodona was unable to push the edits to the processed exercises. The exact error was: %{error}
```
Comment (chvp) [inline config/locales/views/error_mailer/en.yml:19]:
```suggestion
          While processing "%{repository}", Dodona was unable to push the edits to the processed exercises. The exact error was: %{error}.
```

File: config/locales/views/error_mailer/nl.yml [hunk 6]
```diff
@@ -10,3 +10,15 @@ nl:
         other: 'Bij het automatisch verwerken van de oefeningenrepository "%{repository}" liep iets fout. Vermoedelijk zit er een syntaxfout in het configuratiebestand. Onderstaande foutmeldingen werden gegenereerd:'
       regards: 'Vriendelijke groeten,'
       auto-generated: ps. Dit bericht werd automatisch gegenereerd.
+    git_error:
+      subject: 'Pushen naar "%{repository}" mislukt'
```
Comment (chvp) [inline config/locales/views/error_mailer/nl.yml:14]:
```suggestion
      subject: '[Dodona] Pushen naar "%{repository}" mislukt'
```

File: config/locales/views/error_mailer/en.yml [hunk 7]
```diff
@@ -10,3 +10,15 @@ en:
         other: 'Something went wrong during the processing of the exercise repository "%{repository}". You probably made a syntax error in the config file. These were the generated error messages:'
       regards: "Kind regards,"
       auto-generated: This message was generated automatically.
+    git_error:
+      subject: 'Failed to push to "%{repository}"'
```
Comment (chvp) [inline config/locales/views/error_mailer/en.yml:14]:
```suggestion
      subject: '[Dodona] Failed to push to "%{repository}"'
```

File: config/locales/views/error_mailer/nl.yml [hunk 8]
```diff
@@ -10,3 +10,15 @@ nl:
...
+      body:
+        greeting: |
+          Beste %{name},
+        error_message_html: |
```
Comment (chvp) [inline config/locales/views/error_mailer/nl.yml:18]: The `_html` postfix is used for translations that include HTML which should not be escaped. That's not the case for this translation, so it should be left out.

File: app/mailers/error_mailer.rb [hunk 9]
```diff
@@ -23,4 +26,30 @@ def json_error(error, user: nil, name: nil, email: nil)
...
+             name: @name
+           ) + I18n.t(
+             'error_mailer.git_error.body.error_message_html',
+             repository: error.repository.name,
+             error: error.errorstring
+           ) + I18n.t(
+             'error_mailer.git_error.body.regards'
+           ) + I18n.t(
+             'error_mailer.git_error.body.auto-generated'
+           )
```
Comment (chvp) [inline app/mailers/error_mailer.rb:52]: Did you have a reason not to create a view in `app/views/error_mailer`? Then we can serve a multipart email with both HTML and plain text.

File: test/models/repository_test.rb [hunk 10]
```diff
@@ -304,4 +304,35 @@ def teardown
...
+    @remote.commit('copy exercise')
+
+    # should raise DodonaGitError because commit fails
+    @repository.reset
+    assert_raises(DodonaGitError) do
+      @repository.process_activities
+    end
+  end
+
+  test 'should send a mail chen commit fails' do
```
Comment (chvp) [inline test/models/repository_test.rb:324]:
```suggestion
  test 'should send a mail when commit fails' do
```

File: app/mailers/error_mailer.rb (hunk 9 above)
Comment (BramDevlaminck) [reply app/mailers/error_mailer.rb:52]: No I did not have a reason to do this. I just did not notice, I'll change this

File: config/locales/views/error_mailer/nl.yml (hunk 8 above)
Comment (BramDevlaminck) [reply config/locales/views/error_mailer/nl.yml]: it seems this is needed for 1 link that is in the message:

without `_html` suffix
image

with `_html` suffix
image

Comment (chvp) [reply config/locales/views/error_mailer/nl.yml]: Where does this link come from? Only the name is set [here](https://github.com/dodona-edu/dodona/pull/3860/files/44009a6bb61eec67b52a1e95e0ba49f8814f506d#diff-e45dbb23b29d2ff475462fffb81b0fa06f75f18281dca9bb9e3b37aec00424deR46) as far as I can tell?

Comment (jorg-vr) [reply config/locales/views/error_mailer/nl.yml]: `The exact error was: %{error}.` It is in the errorstring

Comment (jorg-vr) [reply config/locales/views/error_mailer/nl.yml]: ah but is the name taht turns into a link, my mistake, odd

Comment (BramDevlaminck) [reply config/locales/views/error_mailer/nl.yml]: I found my mistake. I forgot a `.name` in the `.html.erb` file

**Review by chvp — APPROVED** [review]

File: config/locales/views/error_mailer/nl.yml [hunk 11]
```diff
@@ -10,3 +10,10 @@ nl:
         other: 'Bij het automatisch verwerken van de oefeningenrepository "%{repository}" liep iets fout. Vermoedelijk zit er een syntaxfout in het configuratiebestand. Onderstaande foutmeldingen werden gegenereerd:'
       regards: 'Vriendelijke groeten,'
       auto-generated: ps. Dit bericht werd automatisch gegenereerd.
+    git_error:
+      subject: '[Dodona] Pushen naar "%{repository}" mislukt'
+      greeting: 'Beste %{name},'
+      error_message: 'Tijdens het verwerken van "%{repository}" kon Dodona de aanpassingen aan de verwerkte oefeningen niet pushen. De exacte error was: %{error}.'
+      hint: 'Zorg er zeker voor dat de Dodona gebruiker pushrechten heeft op je repository.'
+      regards: 'Vriendelijke groeten,'
+      auto-generated: ps. Dit bericht werd automatisch gegenereerd.
```
Comment (niknetniko) [inline config/locales/views/error_mailer/nl.yml:19]:
```suggestion
      auto-generated: PS Dit bericht werd automatisch gegenereerd.
```

File: config/locales/views/error_mailer/nl.yml [hunk 12]
```diff
@@ -10,3 +10,10 @@ nl:
         other: 'Bij het automatisch verwerken van de oefeningenrepository "%{repository}" liep iets fout. Vermoedelijk zit er een syntaxfout in het configuratiebestand. Onderstaande foutmeldingen werden gegenereerd:'
       regards: 'Vriendelijke groeten,'
       auto-generated: ps. Dit bericht werd automatisch gegenereerd.
+    git_error:
+      subject: '[Dodona] Pushen naar "%{repository}" mislukt'
+      greeting: 'Beste %{name},'
+      error_message: 'Tijdens het verwerken van "%{repository}" kon Dodona de aanpassingen aan de verwerkte oefeningen niet pushen. De exacte error was: %{error}.'
+      hint: 'Zorg er zeker voor dat de Dodona gebruiker pushrechten heeft op je repository.'
```
Comment (niknetniko) [inline config/locales/views/error_mailer/nl.yml:17]:
```suggestion
      hint: 'Zorg er zeker voor dat de Dodona-gebruiker pushrechten heeft op je repository.'
```

**Review by niknetniko — CHANGES_REQUESTED** [review]

**Review by niknetniko — APPROVED** [review]

**Review by jorg-vr — APPROVED** [review]

**Pull request merged**